### ISSUE TITLE
ESM loader in v11 does not work with got package, after installing the latest import-in-the-middle

### ISSUE THREAD

**Issue by corevo** [newrelic/node-newrelic#1920]
[NOTE]: # ( ^^ Provide a general summary of the issue in the title above. ^^ )

## Description

Attempting to use v11 in an ESM project that includes got causes the node process to crash.

In Node v18 you will get a silent exit code 13. In Node v20, it will throw a SyntaxError before exit code 1.

```
app-1  | (node:8) ExperimentalWarning: `--experimental-loader` may be removed in the future; instead use `register()`:
app-1  | --import 'data:text/javascript,import { register } from "node:module"; import { pathToFileURL } from "node:url"; register("newrelic/esm-loader.mjs", pathToFileURL("./"));'
app-1  | (Use `node --trace-warnings ...` to show where the warning was created)
app-1  | file:///app/node_modules/@payzen/got-commons/node_modules/got/dist/source/index.js?iitm=true:39
app-1  |     let $default = _.default
app-1  |         ^
app-1  |
app-1  | SyntaxError: Identifier '$default' has already been declared
app-1  |     at ModuleLoader.moduleStrategy (node:internal/modules/esm/translators:155:18)
app-1  |     at callTranslator (node:internal/modules/esm/loader:285:14)
app-1  |     at ModuleLoader.moduleProvider (node:internal/modules/esm/loader:291:30)
app-1  |
app-1  | Node.js v20.10.0
app-1 exited with code 1
```

[NOTE]: # ( Describe the problem you're encountering. )
[TIP]:  # ( Do NOT share sensitive information, whether personal, proprietary, or otherwise! )

## Expected Behavior

Running `node --loader newrelic/esm-loader.mjs` should work without crashing node.
[NOTE]: # ( Tell us what you expected to happen. )

## [Troubleshooting](https://discuss.newrelic.com/t/troubleshooting-frameworks/108787) or [NR Diag](https://docs.newrelic.com/docs/using-new-relic/cross-product-functions/troubleshooting/new-relic-diagnostics) results

[NOTE]: # ( Provide any other relevant log data. )
[TIP]:  # ( Scrub logs and diagnostic information for sensitive information )

## Steps to Reproduce

1. run `npm init -y`
2. run `npm i --save got newrelic`
3. create the file index.mjs with the contents
```
import got from 'got'
```
4. run `node --loader newrelic/esm-loader.mjs index.mjs`
5. crash

[NOTE]: # ( Please be as specific as possible. )
[TIP]:  # ( Link a sample application that demonstrates the issue. )

## Your Environment

[TIP]:  # ( Include as many relevant details about your environment as possible. )

* ex: Browser name and version: not relevant
* ex: Node version: 20.10.0
* ex: Operating System and version: Linux

## Additional context

[TIP]:  # ( Add any other context about the problem here. For example, relevant community posts or support tickets. )

**Comment by jsumners-nr** [comment]
What version of the agent are you using? What version of `import-in-the-middle` is being used? The output of `npm ls newrelic import-in-the-middle` should help us proceed.

**Comment by zintus** [comment]
https://github.com/zintus/newrelic-node-examples/commit/c209a0dcc7bebada041a6fcca72a7b771fd0d516
I believe this repro is related. Please advise

`npm run start` fails
running `node index.js` succeeds

Node v21.3.0

**Comment by corevo** [comment]
```
  └─┬ newrelic@11.7.0
    └── import-in-the-middle@1.7.2
```

**Comment by jsumners-nr** [comment]
Thank you. I have narrowed down the issue and submitted a PR upstream (https://github.com/DataDog/import-in-the-middle/pull/53).

**Comment by ljharb** [comment]
You can't export default multiple times, so it's possible this is a node loader bug.

**Comment by jsumners-nr** [comment]
Thank you @ljharb. It's a problem with IITM. I wrote a test for core and it passes:

<details>
<summary>Core Test</summary>

```diff
diff --git a/test/es-module/test-esm-initialization.mjs b/test/es-module/test-esm-initialization.mjs
index f03a47d5d3..e7882e0839 100644
--- a/test/es-module/test-esm-initialization.mjs
+++ b/test/es-module/test-esm-initialization.mjs
@@ -28,3 +28,19 @@ describe('ESM: ensure initialization happens only once', { concurrency: true },
     assert.strictEqual(code, 0);
   });
 });
+
+describe('ESM: nested default exports are handled correctly', () => {
+  it(async() => {
+    const { code, stderr, stdout } = await spawnPromisified(execPath, [
+      '--experimental-import-meta-resolve',
+      '--loader',
+      fixtures.fileURL('es-module-loaders', 'loader-resolve-passthru.mjs'),
+      '--no-warnings',
+      fixtures.path('es-modules', 'nested-default', 'index.mjs'),
+    ]);
+
+    assert.strictEqual(stderr, '');
+    assert.strictEqual(stdout.match(/resolve passthru/g)?.length, 3);
+    assert.strictEqual(code, 0);
+  });
+});
diff --git a/test/fixtures/es-modules/nested-default/a.mjs b/test/fixtures/es-modules/nested-default/a.mjs
new file mode 100644
index 0000000000..dc07e2033d
--- /dev/null
+++ b/test/fixtures/es-modules/nested-default/a.mjs
@@ -0,0 +1,3 @@
+const a = 'a'
+export default a
+export * from './b.mjs'
diff --git a/test/fixtures/es-modules/nested-default/b.mjs b/test/fixtures/es-modules/nested-default/b.mjs
new file mode 100644
index 0000000000..7f5dadf656
--- /dev/null
+++ b/test/fixtures/es-modules/nested-default/b.mjs
@@ -0,0 +1,2 @@
+const b = 'b'
+export default b
diff --git a/test/fixtures/es-modules/nested-default/index.mjs b/test/fixtures/es-modules/nested-default/index.mjs
new file mode 100644
index 0000000000..755d201fc8
--- /dev/null
+++ b/test/fixtures/es-modules/nested-default/index.mjs
@@ -0,0 +1 @@
+export * from './a.mjs'
```

</details>

**Comment by zintus** [comment]
I can confirm your fix helps with my repro too. Thanks for the investigation @jsumners-nr

**Comment by corevo** [comment]
I've tried your fix and the error is now different, it looks like the module loading is okay, but not all the exports are re-exported correctly, it looks like the exported object is got's `Request` rather than the got object itself, I'm using `got@14.0.0` which is a native ESM module.

```
TypeError: got.extend is not a function
    at file:///app/node_modules/@payzen/got-commons/build/got-commons.js:185:29
    at ModuleJob.run (node:internal/modules/esm/module_job:218:25)
    at async ModuleLoader.import (node:internal/modules/esm/loader:329:24)
    at async loadESM (node:internal/process/esm_loader:28:7)
    at async handleMainPromise (node:internal/modules/run_main:120:12)
```

Printing the default exports yields
```
[class Request extends Duplex]
```

**Comment by jsumners-nr** [comment]
Thank you for the update. Figuring out the full solution given that information will be tricky indeed.

**Comment by jsumners-nr** [comment]
I have determined the underlying problem. It is going to require some significant work in IITM.

**Comment by jsumners-nr** [comment]
The issue is that in https://github.com/DataDog/import-in-the-middle/blob/ae292fb72d2a3444f3c4fa77439be81df49ff8f9/hook.js#L205-L212 we merge all transitive exports into a singular namespace. Whatever was the last "default" export in derived namespaces ends up being the default exported from the shim module. Which means we need to figure out how to provide all of the required exports from the shim module while also picking the primary default and renaming the others according to their module name (e.g. './foo-bar.js' default export should be renamed and exported as `fooBar`).

I'm at a loss for how to accomplish this. I welcome any ideas.

**Comment by corevo** [comment]
Sorry but I think I don't fully understand the issue here.

Using the keyword `default` more than once is illegal, it will return a parsing error of duplicate definition of `.default` on the module object, are you referring to a module like this

```js
/* first.mjs */
export default 1

/* second.mjs */
export default 2

/* index.mjs */
export * from './first.mjs'
export * from './second.mjs'
```

Because if so then in `index.mjs` the `.default` is not defined because the default keyword was not strictly used and the exported module will be `{}`, unless you want the default and then you have to rename it if imported from more than one module.

**Comment by ljharb** [comment]
export * from doesn’t bring the default export, so there shouldn’t ever be more than one.

**Comment by jsumners-nr** [comment]
After my last update I had a brainwave and seem to have a viable solution. I'm currently working through cleaning it up and expanding my test to try and cover all bases.

**Comment by jsumners-nr** [comment]
I believe commit https://github.com/DataDog/import-in-the-middle/pull/53/commits/1abc01046de8fca3043e9647f565576dbffbf03a solves the problem.

**Comment by jsumners-nr** [comment]
I think we should see a release soon, but in the meantime, if you'd like to try the latest main branch of `import-in-the-middle` it should solve the issue.

**Comment by alexey-yarmosh** [comment]
Hey, thanks for the fix. I've just checked with the new version of `import-in-the-middle`. It fixes `got` issue.

In the meantime it still fails with another import in our project. When I try to:
```
import { defineScript } from 'redis';
```
I get:
```
node --loader newrelic/esm-loader.mjs test.mjs
(node:97636) ExperimentalWarning: `--experimental-loader` may be removed in the future; instead use `register()`:
--import 'data:text/javascript,import { register } from "node:module"; import { pathToFileURL } from "node:url"; register("newrelic/esm-loader.mjs", pathToFileURL("./"));'
(Use `node --trace-warnings ...` to show where the warning was created)
import { defineScript } from 'redis';
         ^^^^^^^^^^^^
SyntaxError: The requested module 'redis' does not provide an export named 'defineScript'
    at ModuleJob._instantiate (node:internal/modules/esm/module_job:132:21)
    at async ModuleJob.run (node:internal/modules/esm/module_job:214:5)
    at async ModuleLoader.import (node:internal/modules/esm/loader:329:24)
    at async loadESM (node:internal/process/esm_loader:34:7)
    at async handleMainPromise (node:internal/modules/run_main:113:12)

Node.js v20.10.0
```

Also seems problem is specificly with `defineScript`, cause `import { createClient } from 'redis';` works fine.

**Comment by barryhagan** [comment]
> 
> In the meantime it still fails with another import in our project. When I try to:
>...
> Also seems problem is specificly with `defineScript`, cause `import { createClient } from 'redis';` works fine.

This is because of the [re-export issue](https://github.com/DataDog/import-in-the-middle/issues/29).  Only `defineScript` is a re-export, not `createClient`.

We've been waiting for some time for this to be able to move to version 11 and it looks like they finally merged the 6 month old PR.

**Comment by jsumners-nr** [comment]
IITM 1.7.3 has been published and resolves the original issue described here. Please update your dependencies to get the fix. If there are other problems, please open a new issue.

**Comment by alexey-yarmosh** [comment]
Hey, do you have plans/approximate dates to update iitm version in node-newrelic itself?

**Comment by jsumners-nr** [comment]
The update is covered by the semver range qualifier specified in the `package.json`:

https://github.com/newrelic/node-newrelic/blob/f414564738ab3ad8df207f96a857f27309d01f0b/package.json#L203

Updating your dependencies will net the fix.

**Comment by alexey-yarmosh** [comment]
Thanks!

**Comment by zintus** [comment]
Thanks for working on this! @jsumners-nr 

My repro i've linked before still fails with same steps. Please advise
https://github.com/newrelic/newrelic-node-examples/compare/main...zintus:newrelic-node-examples:main

```

> esm-app@1.0.0 start
> node --experimental-loader newrelic/esm-loader.mjs -r newrelic index.js

(node:36337) ExperimentalWarning: `--experimental-loader` may be removed in the future; instead use `register()`:
--import 'data:text/javascript,import { register } from "node:module"; import { pathToFileURL } from "node:url"; register("newrelic/esm-loader.mjs", pathToFileURL("./"));'
(Use `node --trace-warnings ...` to show where the warning was created)
file:///Users/zintus/workspace/newrelic-node-examples/esm-app/node_modules/openai/resources/index.mjs?iitm=true:55
    let $Completions = namespace.Completions
        ^

SyntaxError: Identifier '$Completions' has already been declared
    at ModuleLoader.moduleStrategy (node:internal/modules/esm/translators:167:18)
    at callTranslator (node:internal/modules/esm/loader:285:14)
    at ModuleLoader.moduleProvider (node:internal/modules/esm/loader:291:30)
```

**Comment by jsumners-nr** [comment]
There have been a few new releases of `import-in-the-middle`. Does your issue still persist, @zintus ?
